source/Cosmos.HAL2/Drivers/PCI/Audio/AC97.cs

```diff
@@ -0,0 +1,301 @@
+
+            // Reset PCM out
+            pTransferControl.Byte = (byte)(pTransferControl.Byte | TC_TRANSFER_RESET);
+            while ((pTransferControl.Byte & TC_TRANSFER_RESET) != 0)
+            {
+                // Wait until the byte is cleared
+            }
+
+            // Volume
+            pMasterVolume.Word = CreateMixerVolumeValue(AC97_VOLUME_MAX, AC97_VOLUME_MAX, false);
+            pPCMOutVolume.Word = CreateMixerVolumeValue(AC97_VOLUME_MAX, AC97_VOLUME_MAX, false);
+
+            // Create all needed buffers
+            CreateBuffers(bufferSize);
+
+            // Initialization done - driver can now be activated by using Enable()
+        }
+
+        /// <summary>
+        /// The global instance of the AC97. This property will return
+        /// <see langword="null"/> if the driver has not been initialized.
+        /// </summary>
+        public static AC97 Instance { get; private set; } = null;
+
+        /// <summary>
+        /// Initializes the AC97 driver. This method will return
+        /// an existing instance if the driver is already initialized
+        /// and has a running instance.
+        /// </summary>
+        /// <param name="bufferSize">The buffer size in samples to use. This value cannot be an odd number, as per the AC97 specification.</param>
+        /// <exception cref="ArgumentException">Thrown when the given buffer size is incorrect.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when no AC97-compatible sound card is present.</exception>
+        public static AC97 Initialize(ushort bufferSize)
+        {
+            if (Instance != null)
+                return Instance;
+
+            Instance = new AC97(bufferSize);
+            return Instance;
+        }
+
+        const int BUFFER_COUNT = 32;
+        private unsafe void CreateBuffers(ushort bufferSize)
+        {
+            transferBuffer = new AudioBuffer(bufferSize, new SampleFormat(AudioBitDepth.Bits16, 2, true));
+            bufferSizeSamples = bufferSize;
+            bufferSizeBytes = bufferSize * transferBuffer.format.size;
+
+            bufferDescriptorList = new BufferDescriptorListEntry[BUFFER_COUNT];
+            buffers = new byte[BUFFER_COUNT][];
+
+            for (int i = 0; i < BUFFER_COUNT; i++)
+            {
+                buffers[i] = new byte[bufferSizeBytes];
+                fixed (byte* ptr = buffers[i])
+                {
+                    bufferDescriptorList[i].pointer = ptr;
+                }
+
+                bufferDescriptorList[i].bufferSize = bufferSize;
+                bufferDescriptorList[i].configuration |= BD_FIRE_INTERRUPT_ON_CLEAR;
+            }
+
+            // Tell BDL location
+            fixed (void* ptr = bufferDescriptorList)
+            {
+                pBufferDescriptors.DWord = (uint)ptr;
+            }
+
+            // Set last valid index
+            lastValidIdx = 2;
```

Comment:
I assume the 2 is due to only two audio channels being supported? For clarity, could you please define an enum or const, which makes the meaning of the 2 clear.

Reply:
Honestly, the driver will work on any value that's in the range of <0; 32>... it's more of a remnant when I was testing the driver ^^'>
I'll change this to 0 in a future commit to this PR.